File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main {
    static StringTokenizer st;
    static int[][] triangle = new int[501][501];
    static int n, res = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        for (int r = 1; r <= n; r++) {
            st = new StringTokenizer(br.readLine());

            for (int c = 1; c <= r; c++) {
                triangle[r][c] = nextInt() + Math.max(triangle[r - 1][c - 1], triangle[r - 1][c]);
            }
        }

        for (int c = 1; c <= n; c++) {
            res = Math.max(res, triangle[n][c]);
        }

        System.out.println(res);
        br.close();
    }

    static int nextInt() {
        return Integer.parseInt(st.nextToken());
    }
}

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.kt
import java.io.StreamTokenizer
import kotlin.math.max

fun main() = with(StreamTokenizer(System.`in`.bufferedReader())) {
    val readInt = { nextToken(); nval.toInt() }
    val n = readInt()
    val dp = Array(n + 1) { IntArray(it + 2) }

    repeat(n + 1) { r ->
        repeat(r) { c -> dp[r][c + 1] = readInt() }
    }

    for (r in 2..n) {
        for (c in 1..r) {
            dp[r][c] += max(dp[r - 1][c - 1], dp[r - 1][c])
        }
    }

    with(System.out.bufferedWriter()) {
        append("${dp[n].max()}")
        flush()
        close()
    }
}